File: ai-solutions/03/gemini-3-pro/js/solution.js

const fs = require('fs');
const path = require('path');

function solveLine(line, k) {
    const digits = line.trim().split('').map(Number);
    const n = digits.length;
    let result = [];
    let start = 0;

    for (let i = 0; i < k; i++) {
        // We need to pick one digit.
        // Remaining needed: k - 1 - i
        // End index (exclusive) for search: n - (k - 1 - i) = n - k + i + 1
        let end = n - k + i + 1;
        
        let maxD = -1;
        let maxIdxInWindow = -1;

        // Search window from start to end
        for (let j = start; j < end; j++) {
            if (digits[j] > maxD) {
                maxD = digits[j];
                maxIdxInWindow = j;
            }
            if (maxD === 9) break; // Optimization
        }

        result.push(maxD);
        start = maxIdxInWindow + 1;
    }

    return result.join('');
}

function main() {
    const inputPath = path.join(__dirname, '../../../../inputs/03.txt');
    try {
        const data = fs.readFileSync(inputPath, 'utf8');
        const lines = data.split('\n').filter(line => line.trim() !== '');

        let totalP1 = 0;
        let totalP2 = 0n; // BigInt for part 2

        for (const line of lines) {
            const res1 = solveLine(line, 2);
            totalP1 += parseInt(res1, 10);
            
            const res2 = solveLine(line, 12);
            totalP2 += BigInt(res2);
        }

        console.log(totalP1);
        console.log(totalP2.toString());

    } catch (err) {
        console.error("Error reading file:", err);
    }
}

main();

File: ai-solutions/03/gemini-3-pro/python/solution.py

import os

def solve_line(line, k):
    digits = [int(c) for c in line.strip()]
    n = len(digits)
    result = []
    start = 0
    
    for i in range(k):
        # We need to pick one digit.
        # We need to leave k - 1 - i digits for the remaining steps.
        # So we can search up to n - (k - i)
        # range end is exclusive in python, so + 1
        end = n - (k - i) + 1
        
        best_digit = -1
        best_idx = -1
        
        # Search for the largest digit in the valid range
        # We prefer the earliest occurrence to leave more room (though greedy choice just needs any valid max)
        # Actually, for the greedy strategy to work, we pick the largest digit. 
        # If there are ties, picking the earliest one is always better or equal because it leaves a larger suffix.
        
        # Optimization: no need to scan efficiently if k and n are small enough, but let's be reasonably efficient.
        # slicing digits[start:end]
        
        search_window = digits[start:end]
        max_d = -1
        max_idx_in_window = -1
        
        for idx, d in enumerate(search_window):
            if d > max_d:
                max_d = d
                max_idx_in_window = idx
            if d == 9: # Optimization: 9 is max possible, take it immediately
                break
                
        result.append(max_d)
        start = start + max_idx_in_window + 1
        
    return int("".join(map(str, result)))

def main():
    input_path = os.path.join(os.path.dirname(__file__), "../../../../inputs/03.txt")
    with open(input_path, 'r') as f:
        lines = [line.strip() for line in f if line.strip()]

    total_p1 = 0
    total_p2 = 0
    
    for line in lines:
        total_p1 += solve_line(line, 2)
        total_p2 += solve_line(line, 12)
        
    print(total_p1)
    print(total_p2)

if __name__ == "__main__":
    main()
